src/compositor/windowpixmapitem.cpp

```diff
@@ -480,13 +496,119 @@ QSGNode *WindowPixmapItem::updatePaintNode(QSGNode *oldNode, UpdatePaintNodeData
+        int sizeLocation;
+        int textureLocation;
+        int matrixLocation;
+    };
+
+    QSGTextureProvider *provider = m_item->textureProvider();
+    QSGTexture *texture = provider->texture();
+
+    if (!m_hasBuffer && texture) {
+        if (!m_textureProvider) {
+            SnapshotTextureProvider *prov = new SnapshotTextureProvider;
+            m_textureProvider = prov;
+
+            prov->fbo = 0;
+            prov->program = new QOpenGLShaderProgram;
+            prov->program->addShaderFromSourceCode(QOpenGLShader::Vertex,
+                "attribute highp vec4 vertex;\n"
+                "uniform highp mat4 matrix;\n"
+                "uniform highp vec2 size;\n"
+                "varying highp vec2 texPos;\n"
+                "void main(void) {\n"
+                "   texPos = vertex.xy;\n"
+                "   gl_Position = matrix * vec4(vertex.xy * size, vertex.zw);\n"
+                "}");
+            prov->program->addShaderFromSourceCode(QOpenGLShader::Fragment,
+                "uniform sampler2D texture;\n"
+                "varying highp vec2 texPos;\n"
+                "void main(void) {\n"
+                "   gl_FragColor = texture2D(texture, texPos);\n"
+                "}");
+            if (!prov->program->link())
+                qDebug() << prov->program->log();
+
+            prov->vertexLocation = prov->program->attributeLocation("vertex");
+            prov->sizeLocation = prov->program->uniformLocation("size");
+            prov->textureLocation = prov->program->uniformLocation("texture");
+            prov->matrixLocation = prov->program->uniformLocation("matrix");
+        }
+        provider = m_textureProvider;
+
+        if (m_unmapLock) {
+            SnapshotTextureProvider *prov = static_cast<SnapshotTextureProvider *>(provider);
+
+            if (!prov->fbo || prov->fbo->size() != QSize(width(), height())) {
+                delete prov->fbo;
+                prov->fbo = new QOpenGLFramebufferObject(width(), height());
+            }
+
+            prov->fbo->bind();
+            prov->program->bind();
+
+            glActiveTexture(GL_TEXTURE0);
+            texture->bind();
+            glUniform1i(prov->textureLocation, 0);
+
+            static GLfloat const triangleVertices[] = {
+                1.f, 0.f,
+                1.f, 1.f,
+                0.f, 0.f,
+                0.f, 1.f,
+            };
+            prov->program->enableAttributeArray(prov->vertexLocation);
+            prov->program->setAttributeArray(prov->vertexLocation, triangleVertices, 2);
+
+            QMatrix4x4 pmvMatrix;
+
+            glViewport(0, 0, width(), height());
+            glDisable(GL_BLEND);
+            pmvMatrix.ortho(QRect(0, height(), width(), -height()));
+            prov->program->setUniformValue(prov->matrixLocation, pmvMatrix);
+            prov->program->setUniformValue(prov->sizeLocation, QSizeF(width(), height()));
+            glDrawArrays(GL_TRIANGLE_STRIP, 0, 6);
```

Comment:
It often doesn't.. Chances are good that the triangles you end up creating from the random data on the stack are either nan or way off in the distance outside the viewport. Plus, since the program will fill pixels based on their position on screen only, any such overlapping triangles would just draw the same content again.